lib/Sema/TypeCheckType.cpp

```diff
@@ -1304,7 +1304,12 @@ static SelfTypeKind getSelfTypeKind(DeclContext *dc,
 
   // In enums and structs, 'Self' is just a shorthand for the nominal type,
   // and can be used anywhere.
-  if (!typeDC->getSelfClassDecl())
+  auto *const classDecl = typeDC->getSelfClassDecl();
+  if (!classDecl)
+    return SelfTypeKind::StaticSelf;
+
+  // In a final class, 'Self' is a shorthand for the nominal type.
+  if (classDecl->isFinal())
```

Comment:
Unfortunately I suspect this is an ABI break since it will change the mangling of this method:

```
final class C {
  func myMethod() -> Self { return self }
}
```

Reply:
Argh, well spotted. Let us leave out the covariant result case for now then. 

> I'm not sure we can do this because of the ABI impact, unfortunately.

Would it work if we keep mangling these specific cases as if the `Self` were dynamic? Or will the demangler have the information to figure out the correct type? I suppose this could also be done with a combination of type checker relaxations and sil interceptions, but that seems more cumbersome.